docs: Improve doc about TLS (#1019)
redpanda-data/documentation: docs/manage/kubernetes/security/kubernetes-tls.mdx

    <meta name="description" content="Use TLS to authenticate Redpanda brokers and encrypt communication between clients and brokers."/>
</head>

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';
import ValuesYaml from '../shared/_values-yaml.mdx'

By default, a Redpanda cluster communicates internally and externally over a plain-text network connection, where any malicious party can see all communication.
To secure the communication between Redpanda brokers and clients, configure TLS (Transport Layer Security).

To configure TLS you can [use a self-signed certificate](#use-a-self-signed-certificate) for all listeners.

## Prerequisites

You need the following:

- [cert-manager](https://cert-manager.io/docs/installation/helm/)

  For TLS certificate management, the Redpanda Helm chart relies on cert-manager to simplify the process of obtaining, renewing, and using certificates. cert-manager is a Kubernetes controller that defines custom Kubernetes resource types that make it easier to work with certificates. For details, see the [cert-manager documentation](https://cert-manager.io/docs/).

> -  A custom domain

Reviewer: custom domain that could be internal k8s domain if workloads lives inside the same k8s cluster.

Reviewer: Or maybe it is helm chart current requirement

Reviewer 2: I do not remember this being a requirement? We can enable internal tls no? or am I reading this wrong?

Author: I believe this is required if we want to allow users to access the cluster externally, which seems to be a common use case for our users.

We don't/can't add worker node IP addresses to the SAN list, so any attempt as connecting externally results in:

`unable to request metadata: unable to dial: x509: cannot validate certificate for <IP> because it doesn't contain any IP SANs`

This fix allowed us to use custom domains with self-signed certs to enable external access: https://github.com/redpanda-data/helm-charts/issues/269

- [External access](../../networking/networking-and-connectivity) to your Redpanda cluster.  

  :::note
  Your Redpanda brokers should advertise addresses in your custom domain.
  :::

- [rpk](../../../../get-started/rpk-install) installed on your local machine so that you can test connections to your Redpanda cluster from outside Kubernetes.

## Use a self-signed certificate

A self-signed certificate is signed with its own private key, instead of requesting it from a public or private certificate authority (CA).

1. Enable TLS. The Helm chart generates a self-signed certificate by default.

  Replace the `<custom-domain` placeholder with your domain.

  <Tabs groupId="helm-config">
  <TabItem value="values" label="--values">

  ```yaml title="self-signed-tls.yaml"
  tls:
    enabled: true
  external:

>     domain: <custom-domain>

Reviewer: This is required?

---
title: Configure TLS for Redpanda in Kubernetes
tags:
  - Kubernetes
  - Helm configuration
  - Security
---

<head>
    <meta name="title" content="Configure TLS in Kubernetes | Redpanda Docs"/>
    <meta name="description" content="Use TLS to authenticate Redpanda brokers and encrypt communication between clients and brokers."/>
</head>

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';
import ValuesYaml from '../shared/_values-yaml.mdx'

By default, a Redpanda cluster communicates internally and externally over a plain-text network connection, where any malicious party can see all communication. To secure the communication between Redpanda brokers and clients, configure TLS (Transport Layer Security).

:::note

> For TLS certificate management, the Redpanda Helm chart relies on cert-manager to simplify the process of obtaining, renewing, and using certificates. cert-manager is a Kubernetes controller that defines custom Kubernetes resource types that make it simplier to work with certificates. For more details about cert-manager, see the [cert-manager documentation](https://cert-manager.io/docs/).

Reviewer 3: ```suggestion
For TLS certificate management, the Redpanda Helm chart relies on cert-manager to simplify the process of obtaining, renewing, and using certificates. cert-manager is a Kubernetes controller that defines custom Kubernetes resource types that make it easier to work with certificates. For details, see the [cert-manager documentation](https://cert-manager.io/docs/).
```

:::

To configure TLS you can do one of the following:

- [Use a self-signed certificate](#use-a-self-signed-certificate).

> - [Use intermediate certificate authority](#use-an-intermediate-certificate-authority).

Author: ```suggestion
- [Use an intermediate certificate authority](#use-an-intermediate-certificate-authority).
```

  ```

  <details>
  <summary>
  Example output
  </summary>
  
  ```
  CLUSTER
  =======
  redpanda.19ae8532-c8fa-49ed-8b35-82d74813db3a

  BROKERS
  =======
  ID    HOST                                             PORT
  0*    redpanda-0.redpanda.redpanda.svc.cluster.local.  9093
  1     redpanda-1.redpanda.redpanda.svc.cluster.local.  9093
  2     redpanda-2.redpanda.redpanda.svc.cluster.local.  9093
  ```
  
  </details>

1. Save the root certificate authority (CA) to your local file system outside Kubernetes:

  ```bash
  kubectl -n redpanda get secret redpanda-default-root-certificate -o go-template='{{ index .data "ca.crt" | base64decode }}' > ca.crt
  ```

1. Use that root CA to validate your external connection to Redpanda with rpk.

  Replace `<custom-domain>` with your domain. Replace `<external-port>` with the port on which your cluster is exposed.

  ```bash
  rpk cluster info --brokers <custom-domain>:<external-port> --tls-truststore ca.crt
  ```

## Use an intermediate certificate authority

> The Helm chart can use your own Issuer or ClusterIssuer to generate Certificates that take advantage of an ACME certificate provider such as Let's Encrypt or to use any other intermediate certificate authority (CA).

Reviewer 3: ```suggestion
The Helm chart can use your own Issuer or ClusterIssuer to generate certificates that take advantage of an ACME certificate provider, such as Let's Encrypt, or to use any other intermediate certificate authority (CA).
```

Author: We have a doc convention for capitalizing Kubernetes resources. Here, 'Certificates' refers to the resource in Kubernetes rather than a standard certificate.

---
title: Configure TLS for Redpanda in Kubernetes
tags:
  - Kubernetes
  - Helm configuration
  - Security
---

<head>
    <meta name="title" content="Configure TLS in Kubernetes | Redpanda Docs"/>
    <meta name="description" content="Use TLS to authenticate Redpanda brokers and encrypt communication between clients and brokers."/>
</head>

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';
import ValuesYaml from '../shared/_values-yaml.mdx'
import InstallRpkHomebrew from "../../../get-started/shared/_install-rpk-homebrew.mdx"
import InstallRpkLinux from "../../../get-started/shared/_install-rpk-linux.mdx"

By default, a Redpanda cluster communicates internally and externally over a plain-text network connection, where any malicious party can see all communication.
To secure the communication between Redpanda brokers and clients, configure TLS (Transport Layer Security).

To configure TLS you can [use a self-signed certificate](#use-a-self-signed-certificate) for all listeners.

## Prerequisites

You need the following:

> - [cert-manager](https://cert-manager.io/docs/installation/helm/)

Reviewer 4: If only one item here, you don't need a bullet. When you have more than one, go ahead and use bullets.

    --set listeners.kafka.tls.cert=<certificate-name> \
    --set listeners.http.tls.cert=<certificate-name> \
    --set listeners.schemaRegistry.tls.cert=<certificate-name>
  ```

  </TabItem>
  </Tabs>

  :::tip
  <ValuesYaml path="tls"/>
  :::

1. Make sure the Certificates are in a `READY` state.

  ```bash
  kubectl get certificate -n redpanda
  ```

  ```
  NAME                                 READY   SECRET                               AGE
  redpanda-default-cert                True    redpanda-default-cert                10m
  redpanda-default-root-certificate    True    redpanda-default-root-certificate    10m
  redpanda-<certificate-name>-cert     True    redpanda-<certificate-name>-cert     10m
  ```

### Test external connections

Validate your external connection to Redpanda with rpk.

Replace the following placeholders:

- `<subdomain>`: The subdomain that's in the advertised address of one of your Redpanda brokers.
- `<custom-domain>`: Your domain.
- `<external-port>`: The port on which your cluster is exposed.

```bash
rpk cluster info --brokers <subdomain>.<custom-domain>:<external-port>
```

## Troubleshooting

### cannot validate certificate for 127.0.0.1

> You'll see this error if you are using a CA certificate and you try to establish an internal connection using localhost. For example:

Reviewer 4: ```suggestion
This error appears if you are using a CA certificate when you try to establish an internal connection using localhost. For example:
```

```bash
kubectl exec redpanda-0 -c redpanda -n redpanda -- rpk cluster info --tls-enabled
```

Result:

```
unable to request metadata: unable to dial: x509: cannot validate certificate for 127.0.0.1 because it doesn't contain any IP SANs
```

#### Solution

You must specify the public domain:

```
kubectl exec redpanda-0 -c redpanda -n redpanda -- rpk cluster info --brokers <subdomain>.<domain>:<external-port> --tls-enabled
```

Or, [use self-signed certificates](#use-a-self-signed-certificate).

### invalid large response size

> You'll see this error if you don't specify that you are connecting over TLS. For example:

Reviewer 4: ```suggestion
This error appears when you don't specify that you are connecting over TLS. For example:
```
